File: ex/uri1610.java
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;
import java.util.BitSet;
import java.util.List;
import java.util.LinkedList;

public class uri1610 {
  static InputStreamReader ir = new InputStreamReader(System.in);
  static BufferedReader in = new BufferedReader(ir);

  public static final void main(String[] args) throws IOException {
    int numberCases = Integer.parseInt(in.readLine());
    BitSet path;
    List queue;
    String[] line;
    int dom, sub;
    int doc, fil;
    int docs, rels;
    boolean[][] ships;
    boolean outlet;

    for (int c = 0; c < numberCases; c++) {
      /* read data */
      line = in.readLine().split(" ");
      docs = Integer.parseInt(line[0]);
      rels = Integer.parseInt(line[1]);
      ships = new boolean[docs][docs];

      for (int r = 0; r < rels; ++r) {
        line = in.readLine().split(" ");
        dom = Integer.parseInt(line[0]) - 1;
        sub = Integer.parseInt(line[1]) - 1;
        ships[dom][sub] = true;
      }

      /* process data */
      outlet = false;
      for (dom = 0; dom < docs && outlet == false; ++dom) {
        queue = new LinkedList<Integer>();
        path = new BitSet();

        queue.add(dom);
        while (!queue.isEmpty() && outlet == false) {
          dom = (int) queue.remove(0);

          for (sub = 0; sub < docs && outlet == false; ++sub) {
            if (ships[dom][sub] == true && path.get(sub) == false) {
              path.set(sub, true);
              queue.add(sub);
            }
            else if (ships[dom][sub] == true && path.get(sub) == true) {
              outlet = true;
            }
          }
          
        }
      }

      /* write data */
      if (outlet)
        System.out.println("SIM");
      else
        System.out.println("NAO");
    }
  }
}

File: ex/uri1610.cpp
#include <iostream>
#include <stdlib.h>
using namespace std;

bool DFS(bool** graph, int limit, bool* relax, int vertex)
{
	bool result = false;

	if (relax[vertex] == true)
		result = true;

	relax[vertex] = true;
	for (int neigh = 0; neigh < limit && result == false; ++neigh)
		if (graph[vertex][neigh] == true && relax[neigh] == false)
			result = DFS(graph, limit, relax, neigh);
		else if (graph[vertex][neigh] == true && relax[neigh] == true)
			result = true;

	return result;
}

main()
{
	int nc;
	int docs, rels;
	int dom, sub;
	bool outlet;
	bool* path;
	bool** ships;

	cin >> nc;
	for (int c = 0; c < nc; c++)
	{
		cin >> docs >> rels;
		path = (bool*) malloc(sizeof(bool)*docs);
		ships = (bool**) malloc(docs*sizeof(bool*));
		for (dom = 0; dom < docs; ++dom)
		{
			ships[dom] = (bool*) malloc(docs*sizeof(bool));
			for (sub = 0; sub < docs; ++sub)
				ships[dom][sub] = false;
		}

		for (int r = 0; r < rels; ++r)
		{
			cin >> dom >> sub;
			ships[dom-1][sub-1] = true;
		}

		outlet = false;
		/* DFS */
		for (dom = 0; dom < docs && outlet == false; ++dom)
		{
			for (sub = 0; sub < docs; ++sub)
				path[sub] = false;
			outlet = DFS(ships, docs, path, dom);
		}

		if (outlet) cout << "SIM";
		else cout << "NAO";
		cout << endl;
	}
}
